components/net/indexeddb/engines/sqlite.rs

```diff
@@ -0,0 +1,513 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+use std::path::{Path, PathBuf};
+
+use net_traits::indexeddb_thread::{
+    AsyncOperation, AsyncReadOnlyOperation, AsyncReadWriteOperation, CreateObjectResult,
+    IndexedDBKeyRange, KeyPath, PutItemResult,
+};
+use sea_orm::prelude::*;
+use sea_orm::sea_query::IntoCondition;
+use sea_orm::{Condition, Database, IntoActiveModel, NotSet, Set};
+use tokio::sync::oneshot;
+
+use crate::async_runtime::HANDLE;
+use crate::indexeddb::engines::{KvsEngine, KvsTransaction, SanitizedName};
+use crate::indexeddb::idb_thread::IndexedDBDescription;
+
+mod database_model;
+mod object_data_model;
+mod object_store_index_model;
+mod object_store_model;
+// mod serialize;
+mod serialize {
+    use net_traits::indexeddb_thread::IndexedDBKeyType;
+
+    pub fn encode(key: &IndexedDBKeyType) -> Vec<u8> {
+        bincode::serialize(&key).unwrap()
+    }
+
+    pub fn decode(s: &[u8]) -> Option<IndexedDBKeyType> {
+        bincode::deserialize(s).ok()
+    }
+}
+
+// These pragmas need to be set once
+const DB_INIT_PRAGMAS: [&str; 2] = ["PRAGMA journal_mode = WAL;", "PRAGMA encoding = 'UTF-16';"];
+
+// These pragmas need to be run once a connection.
+const DB_PRAGMAS: [&str; 4] = [
+    "PRAGMA synchronous = NORMAL;",
+    "PRAGMA journal_size_limit = 67108864 -- 64 megabytes;",
+    "PRAGMA mmap_size = 67108864 -- 64 megabytes;",
+    "PRAGMA cache_size = 2000;",
+];
+
+fn range_to_query(range: IndexedDBKeyRange) -> Condition {
+    // Special case for optimization
+    if let Some(singleton) = range.as_singleton() {
+        let encoded = serialize::encode(singleton);
+        return object_data_model::Column::Data.eq(encoded).into_condition();
+    }
+    let mut parts = vec![];
+    if let Some(upper) = range.upper.as_ref() {
+        let upper_bytes = serialize::encode(upper);
+        let query = if range.upper_open {
+            object_data_model::Column::Data.lt(upper_bytes)
+        } else {
+            object_data_model::Column::Data.lte(upper_bytes)
+        };
+        parts.push(query);
+    }
+    if let Some(lower) = range.lower.as_ref() {
+        let lower_bytes = serialize::encode(lower);
+        let query = if range.upper_open {
+            object_data_model::Column::Data.gt(lower_bytes)
+        } else {
+            object_data_model::Column::Data.gte(lower_bytes)
+        };
+        parts.push(query);
+    }
+    let mut condition = Condition::all();
+    for part in parts {
+        condition = condition.add(part);
+    }
+    condition
+}
+
+macro_rules! err {
+    ($e:expr) => {
+        Err(format!("{:?}", $e))
+    };
+}
```

Comment:
Same comments as the other err! macro in this PR. I don't think it provides enough value to keep it.